File: Top Interview 150/Binary Tree General/0105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def splitTree(self,preorder, inorder_dic, root_idx, left, right):
        
        # with the root index create first root/subroot
        root = TreeNode(preorder[root_idx])

        #  get mid value
        mid = inorder_dic[preorder[root_idx]]

        #  get left Nodes if exists
        if mid> left:
            root.left = self.splitTree(preorder, inorder_dic, root_idx+1, left,mid -1)
        #  get right nodes if exists
        if mid<right:
            root.right = self.splitTree(preorder, inorder_dic, root_idx+mid - left +1, mid+1, right)
        return root

    def buildTree(self, preorder, inorder):
        inorder_dic = {}
        for i in range(0, len(inorder)):
            inorder_dic[inorder[i]] = i
        return self.splitTree(preorder, inorder_dic, 0,0, len(inorder)-1)
    

File: Top Interview 150/Binary Tree General/0105. Construct Binary Tree from Preorder and Inorder Traversal/105. Construct Binary Tree from Preorder and Inorder Traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preOrder, int[] inOrder) {
        Map<Integer,  Integer> inOrderIndexMap = new HashMap<>();
        // put values in hashmap
        for(int i = 0; i < inOrder.length; i++)
            inOrderIndexMap.put(inOrder[i],i);
        return splitTree(preOrder,inOrderIndexMap, 0, 0,inOrder.length -1);
    }

    public TreeNode splitTree (int[] preOrder, Map<Integer, Integer> inOrderIndexMap, int rootIndex, int left, int right){
        TreeNode root = new TreeNode(preOrder[rootIndex]);

        //create left and right subtree
        // get middle value
        int mid = inOrderIndexMap.get(preOrder[rootIndex]);

        //
        if(mid > left){
            root.left = splitTree(preOrder, inOrderIndexMap,  rootIndex +1, left, mid-1);
        }
        if (mid < right){
            root.right = splitTree(preOrder, inOrderIndexMap, rootIndex + mid - left +1, mid+1, right);
        }
        return root;
    }
}